plutus-tx-plugin/src/PlutusTx/Compiler/Expr.hs

```diff
@@ -848,6 +857,280 @@ compileExpr e = withContextM 2 (sdToTxt $ "Compiling expr:" GHC.<+> GHC.ppr e) $
         GHC.Type _ -> throwPlain $ UnsupportedError "Types as standalone expressions"
         GHC.Coercion _ -> throwPlain $ UnsupportedError "Coercions as expressions"
 
+-- | Checks whether the given application should be compiled in a specific way.
+--
+-- If yes, returns @Just compilation_action@. Otherwise returns @Nothing@, and it will
+-- be compiled the normal way.
+isKnownApp ::
+    CompilingDefault uni fun m ann =>
+    GHC.Var ->
+    [GHC.CoreExpr] ->
+    Maybe (m (PIRTerm uni fun))
+isKnownApp fun args = Map.lookup (splitGhcName (GHC.varName fun)) knownApps >>= ($ args)
+
+-- | A list of applications that should be compiled in specific ways.
+knownApps ::
+    CompilingDefault uni fun m ann =>
+    Map (Maybe String, String) ([GHC.CoreExpr] -> Maybe (m (PIRTerm uni fun)))
+knownApps =
+    Map.fromListWithKey (\n -> error ("knownApps: key defined more than once: " <> show n))
+        . fmap (first splitThName)
+        $ [ ( '(GHC.Classes.==)
+          , \case
+                [GHC.Type ty, _eqDict]
+                    | ty `GHC.eqType` GHC.integerTy ->
+                        -- Compile `(==) @Integer` via `Builtins.equalsInteger`.
```

Comment:
Isn't this pretty redundant with the class-dictionary substitution you do below? You're already going to replace the dictionary for `Eq @Integer`. And if we have `(==) @Integer` then we _do_ have a concrete, known dictionary and we _should_ be able to optimize that away. So I'm unconvinced that we need this for the typeclass functions.